### PR TITLE
Fix #81 map data race

### LINKED ISSUE #81: debug info: concurrent map read and map write
time="2016-06-29T20:34:30+08:00" level=info msg="Metrics persisted to 'persistence.file'." source="diskmetricstore.go:156" 
fatal error: concurrent map read and map write

goroutine 268718 [running]:
runtime.throw(0xa1a5c0, 0x21)
	/usr/local/go/src/runtime/panic.go:547 +0x90 fp=0xc820d21550 sp=0xc820d21538
runtime.mapaccess2(0x8ab2c0, 0xc82235d9e0, 0xc825baa7c0, 0x4e7dca, 0x58c000)
	/usr/local/go/src/runtime/hashmap.go:343 +0x5a fp=0xc820d21598 sp=0xc820d21550
reflect.mapaccess(0x8ab2c0, 0xc82235d9e0, 0xc825baa7c0, 0xc82235d9e0)
	/usr/local/go/src/runtime/hashmap.go:993 +0x35 fp=0xc820d215c8 sp=0xc820d21598
reflect.Value.MapIndex(0x8ab2c0, 0xc825baa398, 0x95, 0x821a60, 0xc825baa7c0, 0x98, 0x0, 0x0, 0x0)
	/usr/local/go/src/reflect/value.go:1041 +0x14a fp=0xc820d21650 sp=0xc820d215c8
encoding/gob.(*Encoder).encodeMap(0xc820092000, 0xc820092038, 0x8ab2c0, 0xc825baa398, 0x95, 0xa8ab98, 0xc821490110, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encode.go:381 +0x2cb fp=0xc820d21760 sp=0xc820d21650
encoding/gob.encOpFor.func3(0xc8200d0f30, 0xc8250e9580, 0x8ab2c0, 0xc825baa398, 0x95)
	/usr/local/go/src/encoding/gob/encode.go:577 +0x10a fp=0xc820d217d0 sp=0xc820d21760
encoding/gob.(*Encoder).encodeStruct(0xc820092000, 0xc820092038, 0xc82216d3e0, 0x920ea0, 0xc825baa390, 0x99)
	/usr/local/go/src/encoding/gob/encode.go:334 +0x3e6 fp=0xc820d218b0 sp=0xc820d217d0
encoding/gob.encOpFor.func4(0x0, 0xc8250e9540, 0x920ea0, 0xc825baa390, 0x99)
	/usr/local/go/src/encoding/gob/encode.go:587 +0xb3 fp=0xc820d21900 sp=0xc820d218b0
encoding/gob.encodeReflectValue(0xc8250e9540, 0x920ea0, 0xc825baa390, 0x99, 0xc821490140, 0x0)
	/usr/local/go/src/encoding/gob/encode.go:369 +0x147 fp=0xc820d21970 sp=0xc820d21900
encoding/gob.(*Encoder).encodeMap(0xc820092000, 0xc820092038, 0x8ab220, 0xc820129e60, 0x15, 0xa8ac00, 0xc821490140, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encode.go:381 +0x329 fp=0xc820d21a80 sp=0xc820d21970
encoding/gob.encOpFor.func3(0xc8238b9e30, 0xc8250e94c0, 0x8ab220, 0xc820129e60, 0x15)
	/usr/local/go/src/encoding/gob/encode.go:577 +0x10a fp=0xc820d21af0 sp=0xc820d21a80
encoding/gob.(*Encoder).encodeSingle(0xc820092000, 0xc820092038, 0xc82216d3c0, 0x8ab220, 0xc820129e60, 0x15)
	/usr/local/go/src/encoding/gob/encode.go:307 +0x26c fp=0xc820d21b88 sp=0xc820d21af0
encoding/gob.(*Encoder).encode(0xc820092000, 0xc820092038, 0x8ab220, 0xc820129e60, 0x15, 0xc821035fc0)
	/usr/local/go/src/encoding/gob/encode.go:709 +0x216 fp=0xc820d21be8 sp=0xc820d21b88
encoding/gob.(*Encoder).EncodeValue(0xc820092000, 0x8ab220, 0xc820129e60, 0x15, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encoder.go:247 +0x6ed fp=0xc820d21d58 sp=0xc820d21be8
encoding/gob.(*Encoder).Encode(0xc820092000, 0x8ab220, 0xc820129e60, 0x0, 0x0)
	/usr/local/go/src/encoding/gob/encoder.go:174 +0x72 fp=0xc820d21da8 sp=0xc820d21d58
github.com/prometheus/pushgateway/storage.(*DiskMetricStore).persist(0xc82014cc30, 0x0, 0x0)
	/go/src/github.com/prometheus/pushgateway/storage/diskmetricstore.go:249 +0x21d fp=0xc820d21ea8 sp=0xc820d21da8
github.com/prometheus/pushgateway/storage.(*DiskMetricStore).loop.func1.1()
	/go/src/github.com/prometheus/pushgateway/storage/diskmetricstore.go:150 +0x7d fp=0xc820d21f90 sp=0xc820d21ea8
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc820d21f98 sp=0xc820d21f90
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [IO wait]:
net.runtime_pollWait(0x7f7aa9ab91a8, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8235a7e20, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8235a7e20, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc8235a7dc0, 0x0, 0x7f7aa9ab9268, 0xc820a48000)
	/usr/local/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(0xc82395a168, 0xc821e12c20, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
net.(*TCPListener).Accept(0xc82395a168, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:264 +0x3d
net/http.(*Server).Serve(0xc820124180, 0x7f7aa9ab8228, 0xc82395a168, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2117 +0x129
main.main()
	/go/src/github.com/prometheus/pushgateway/main.go:108 +0x126a

goroutine 17 [syscall, 50 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x132
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 33 [select]:
github.com/prometheus/pushgateway/storage.(*DiskMetricStore).loop(0xc82014cc30, 0x45d964b800)
	/go/src/github.com/prometheus/pushgateway/storage/diskmetricstore.go:166 +0x4a2
created by github.com/prometheus/pushgateway/storage.NewDiskMetricStore
	/go/src/github.com/prometheus/pushgateway/storage/diskmetricstore.go:79 +0x5a8

goroutine 94 [chan receive, 50 minutes]:
main.interruptHandler(0x7f7aa9ab8228, 0xc82395a168)
	/go/src/github.com/prometheus/pushgateway/main.go:135 +0x192
created by main.main
	/go/src/github.com/prometheus/pushgateway/main.go:107 +0x117b

goroutine 95 [select, 50 minutes, locked to thread]:
runtime.gopark(0xa8bbb0, 0xc82084b728, 0x99e480, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:262 +0x163
runtime.selectgoImpl(0xc82084b728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa67
runtime.selectgo(0xc82084b728)
	/usr/local/go/src/runtime/select.go:215 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:279 +0x358
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 268798 [IO wait]:
net.runtime_pollWait(0x7f7aa9ab8f68, 0x72, 0xc821ab0000)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc820175f70, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820175f70, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820175f10, 0xc821ab0000, 0x1000, 0x1000, 0x0, 0x7f7aa9d4f050, 0xc82000e068)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc821bac780, 0xc821ab0000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.(*connReader).Read(0xc82143c000, 0xc821ab0000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:526 +0x196
bufio.(*Reader).fill(0xc821676420)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc821676420, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*conn).readRequest(0xc82008f600, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:702 +0x2e6
net/http.(*conn).serve(0xc82008f600)
	/usr/local/go/src/net/http/server.go:1425 +0x947
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2137 +0x44e

goroutine 268701 [runnable]:
net/http.(*conn).serve(0xc82175c000)
	/usr/local/go/src/net/http/server.go:1383
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2137 +0x44e

runtime environment：
version: 0.3.0
os: ubuntu 14.04
kernel: 3.2.0-23-generic

### PR DESCRIPTION
**Pull request by slene**

### CODE REVIEW

**Comment by beorn7** [comment]
Thanks for your contribution.

This shouldn't be a race since the only write to dms is triggered by the same one goroutine that also triggers the checkpoint.

Have you seen a race in the race detector?
I'd be interested in the output, if you did.

**Comment by beorn7** [comment]
alright, just saw your issue filed. looking now...

**Comment by beorn7** [comment]
OK, understood now.

File: storage/diskmetricstore.go [hunk 1]
```diff
@@ -246,7 +246,12 @@ func (dms *DiskMetricStore) persist() error {
 	}
 	inProgressFileName := f.Name()
 	e := gob.NewEncoder(f)
-	if err := e.Encode(dms.metricGroups); err != nil {
+
+	if err := func() error {
+		dms.lock.RLock()
+		defer dms.lock.RUnlock()
+		return e.Encode(dms.metricGroups)
+	}(); err != nil {
```
Comment (beorn7) [inline storage/diskmetricstore.go]: It's fine to write it more straight forward:

```go
dms.lock.RLock()
err = e.Encode(dms.metricGroups)
dms.lock.RUnlock()
if err != nil {
// ...
```

err already exists in the scope.

Comment (slene) [inline storage/diskmetricstore.go]: 👌

**Comment by beorn7** [comment]
:+1: Thanks for spotting (and fixing) this.

**Comment by Nodeman123** [comment]
If parameters include '/' ,It will response code 400.How to do?

**Pull request merged**